### ISSUE TITLE
TabItem State Styling requires internal knowledge of control's coding (Bad Design Karma!!!)

### ISSUE THREAD

**Issue by KentSwan** [AvaloniaUI/Avalonia#14130]
## Is your feature request related to a problem? Please describe.

To style the content of a tabitem I have the following  style settings.

		<Style Selector="TabControl">
			<Setter Property="Padding" Value="0 0 0 0"/>
			<Setter Property="Margin" Value="0 0 0 0"/>
		</Style>
		<Style Selector="TabItem">
			<Setter Property="BorderBrush" Value="Blue"/>
			<Setter Property="BorderThickness" Value="2"/>
			<Setter Property="CornerRadius" Value="8 8 0 0"/>
			<Setter Property="Background" Value="White"/>
			<Setter Property="Foreground" Value="Black"/>
		</Style>
		<Style Selector="TabItem:selected /template/ ContentPresenter#PART_ContentPresenter">
			<Setter Property="Background" Value="Black"/>
			<Setter Property="Foreground" Value="Yellow"/>
			<Setter Property="FontWeight" Value="Bold"/>
			<Setter Property="FontStyle" Value="Italic"/>
		</Style>

The first TabItem style sets the unselected tab style elements
The second Tabitem style set the selected tab style elements.

My issue is that the :selected attribute tag should be suffient to do this without having to know and specify control internal information.

## Describe the solution you'd like

When you add state attributes to a control component for styling, the backing concept is that the properties set override the base properties of that control component based on the state attribute being true or ignored if false.

I would thus like to have 
		<Style Selector="TabItem:selected /template/ ContentPresenter#PART_ContentPresenter">
and
		<Style Selector="TabItem:selected">

It is my contention that the two :selector attribute examples should be equivilent and sufficient to map the setter to the control components' internal template,   For long-term regression consistency, all state :attributes should always have a default template property mappings along with an ability to override that mapping by being explicit.  

This conforms to the concept that normally implementation-specific internal knowledge, reference or access should not be required for ordinary use but an override argument syntax is provided when it may be required (typically this would be when using the control as a component part of a more complex control.  In the above example case,  the two forms of the selectors should be functionally identical where the additional mapping, in this case, would be the default internal mapping.

## Describe alternatives you've considered

None as yet. It's irritating to have to go into the control code to obtain the mappings required to get to the state styling properties overloads.

## Additional context
None.

**Comment by jp2masa** [comment]
This has been discussed several times, and it's by design, e.g. https://github.com/AvaloniaUI/Avalonia/pull/2662#issuecomment-515764732.

**Comment by KentSwan** [comment]
So we have to go rooting around in every control to locate the templates that are associated with state-dependent style properties so we can set them?   If this is true then can I assume that "/template/ ContentPresenter#PART_ContentPresenter" will always work in every use case? If not then where is the how-to documentation other than in the rtfc ("Read the F***ing Code.") 

As an aside, I've been doing software design and implementation for more years than I can count.  My comment on that "It's by design" doesn't mean that the design choices are correct or follow accepted object interface practices and their underlying implementations.

Sorry about the soapbox but you hit one of my pet peeves with that "it its what it is" statement.

**Comment by jp2masa** [comment]
Personally, I don't like this design choice either, because it doesn't make sense to set a local value and then see the value changing due to a style. I don't make decisions, though, and the core team has to consider many aspects when deciding. In this case, it was related to WPF compatibility IIRC.

**Comment by thevortexcloud** [comment]
>  every control to locate the templates that are associated

Yes. They are "lookless" controls. Which means they can look like anything. Most people just happen to be using the official themes, but there is nothing stopping somebody else from making it look and visually function completely differently. The only thing controls have to have on them are `PART_` parts. Which you can find by looking at the C# code for the control. EG this is the declaration for the slider: 

https://github.com/AvaloniaUI/Avalonia/blob/38e839997d2c204548e6fad396c178780a010cb1/src/Avalonia.Controls/Slider.cs#L43-L50

But the parts can still be located really anywhere in the control.

**Comment by thevortexcloud** [comment]
> it was related to WPF compatibility IIRC.

This indeed fairly similar to how WPF works and it's also just as annoying trying to find a list of all the control parts.

**Comment by KentSwan** [comment]
Having state-dependent styles is a consistent way to implement visual animation that tells the user what's going on. The issue for an application designer is that they need the styling of things to be relatively consistent throughout the application which means that those design choices may not be what the API defaults to.  In any case, the list of available state : attributes are the same across the board for every control that implements them (which is most of them).  The application UI coder should NEVER have to know anything about how an internal aspect of control is coded or named because the API and Control coders can arbitrarily change it which would break the dependent applications.  Again, this means that this mapping requirement, while currently necessary, is bogus.  

Axaml is already compiled and translated to map both the view and the data context by interpreting the source syntax and mapping it to the implementation.  With the advent of code generators, the API and control maintainers should be able to craft a consistent way of extending the compilation to internally map or accept a template override to the underlying code.  Granted the knowledge required is likely massive but those with experience should be able to make it happen.  

If my suggestion is accepted then future axaml users would be able to code without fussing but having the option to add a different mapping would be regression compliant and thus not break the existing code base.

Question: Does anyone in the Avalonia development team that have design knowledge and implementation control look at the issues we bring up?

**Comment by thevortexcloud** [comment]
> a consistent way of extending the compilation to internally map or accept a template override to the underlying code

XAML can be loaded at runtime, not just at compile time. Control templates can also be defined at runtime. That said I would like it to be possible for compile errors to be raised if a `PART_` is not defined for a template. It would make making/maintaining control templates far less painful.

>  would be regression compliant and thus not break the existing code base.

I fail to see how this is possible. If you are reliant on a theme working in a specific way and then they suddenly change what a pseudo selector does then you now have a breaking change. You are better off asking for the theme author to make their templates use template bindings when possible. `TabItem` for example should already do this for the border control on the fluent theme

https://github.com/AvaloniaUI/Avalonia/blob/38e839997d2c204548e6fad396c178780a010cb1/src/Avalonia.Themes.Fluent/Controls/TabItem.xaml#L30-L36

In which case setting the `Background` property on the `TabItem` would change the background of that `Border` control.
